controller/system_backup_controller.go

```diff
@@ -0,0 +1,1068 @@
+	systemBackupRecordTypeNormal = systemBackupRecordType("normal")
+)
+
+type systemBackupCacheState struct {
+	nextState longhorn.SystemBackupState
+
+	recordType systemBackupRecordType
+	message    string
+	reason     string
+
+	lock *sync.Mutex
+}
+
+type SystemBackupController struct {
+	*baseController
+
+	// which namespace controller is running with
+	namespace string
+
+	// use as the OwnerID of the controller
+	controllerID string
+
+	// the running manager image
+	managerImage string
+
+	kubeClient clientset.Interface
+
+	eventRecorder record.EventRecorder
+
+	ds *datastore.DataStore
+
+	cacheSyncs []cache.InformerSynced
+	cacheState *systemBackupCacheState
+}
+
+func NewSystemBackupController(
+	logger logrus.FieldLogger,
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	kubeClient clientset.Interface,
+	namespace string,
+	controllerID string,
+	managerImage string) *SystemBackupController {
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{
+		Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events(""),
+	})
+
+	c := &SystemBackupController{
+		baseController: newBaseController(SystemBackupControllerName, logger),
+
+		namespace:    namespace,
+		controllerID: controllerID,
+		managerImage: managerImage,
+
+		ds: ds,
+
+		kubeClient:    kubeClient,
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, corev1.EventSource{Component: SystemBackupControllerName + "-controller"}),
+	}
+
+	ds.SystemBackupInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.enqueueSystemBackup,
+		UpdateFunc: func(old, cur interface{}) { c.enqueueSystemBackup(cur) },
+		DeleteFunc: c.enqueueSystemBackup,
+	})
+	c.cacheSyncs = append(c.cacheSyncs, ds.SystemBackupInformer.HasSynced)
+
+	c.cacheState = &systemBackupCacheState{
+		lock: &sync.Mutex{},
+	}
+
+	return c
+}
+
+func (c *SystemBackupController) enqueueSystemBackup(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", obj, err))
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+func (c *SystemBackupController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.logger.Info("Starting Longhorn SystemBackup controller")
+	defer c.logger.Info("Shutting down Longhorn SystemBackup controller")
+
+	if !cache.WaitForNamedCacheSync(c.name, stopCh, c.cacheSyncs...) {
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *SystemBackupController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *SystemBackupController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	defer c.cacheStateReset()
```

Comment:
This looks weird... Have you tried to launch multiple SystemBackup creations on the same node?

Reply:
Thanks, fixed.